### DISCUSSION TITLE
Could `CompileTask` be extended to also include a `noOp` field

### DISCUSSION THREAD

**Discussion by ckipp01** [build-server-protocol/build-server-protocol#654] [Proposals]
The context for this can be found in [here](https://github.com/scalameta/metals/issues/6099#issuecomment-1945960033), but the general idea is that there are times when build servers are sending a `build/taskStart` that is a no-op. I was always under the assumption that this was only known at the end of the compilation cycle and that's why it was included in the `CompileReport`, but in the case of Bloop there seems to be quite a few times where you get a `build/taskStart` where the message is something like _Start no-op compilation for minimal_f4dd477a3a_, so it already knows that this is a no-op. The issue is that without parsing that string there is no way for a client to know this is a no-op causing clients to then forward these as progress notifications to the LSP clients, which display them... but add no value. This would be a change [here](https://github.com/build-server-protocol/build-server-protocol/blob/eac6038263510dad9f299e1bce3a02b6a272763b/spec/src/main/resources/META-INF/smithy/bsp/bsp.smithy#L1210-L1218) to add `noOp`. 

NOTE: That i'm also exploring why we even do this in Bloop.

**Comment by agluszak** [comment]
> The issue is that without parsing that string there is no way for a client to know this is a no-op causing clients to then forward these as progress notifications to the LSP clients, which display them... but add no value. 

Why send them to BSP clients then?

**Reply by ckipp01** [reply]
> NOTE: That i'm also exploring why we even do this in Bloop.

**Reply by ckipp01** [reply]
So after looking into it a bit here is an example:

```
[Trace - 06:01:07 PM] Sending request 'buildTarget/compile - (19)'
Params: {
  "targets": [
    {
      "uri": "file:/Users/ckipp/Documents/scala-workspace/minimal/.scala-build/?id\u003dminimal_f4dd477a3a"
    }
  ],
  "originId": "METALS-$f5924e8b-0e59-4300-8cfc-9173bb99a6b7"
}


[Trace - 06:01:07 PM] Received notification 'build/taskStart'
Params: {
  "taskId": {
    "id": "5"
  },
  "eventTime": 1708016467646,
  "message": "Start no-op compilation for minimal_f4dd477a3a",
  "dataKind": "compile-task",
  "data": {
    "target": {
      "uri": "file:/Users/ckipp/Documents/scala-workspace/minimal/.scala-build/?id\u003dminimal_f4dd477a3a"
    }
  }
}


[Trace - 06:01:07 PM] Received notification 'build/taskFinish'
Params: {
  "taskId": {
    "id": "5"
  },
  "eventTime": 1708016467647,
  "message": "Compiled \u0027minimal_f4dd477a3a\u0027",
  "status": 1,
  "dataKind": "compile-report",
  "data": {
    "target": {
      "uri": "file:/Users/ckipp/Documents/scala-workspace/minimal/.scala-build/?id\u003dminimal_f4dd477a3a"
    },
    "originId": "METALS-$f5924e8b-0e59-4300-8cfc-9173bb99a6b7",
    "errors": 0,
    "warnings": 0,
    "isNoOp": true,
    "isLastCycle": true,
    "clientDir": "file:///Users/ckipp/Documents/scala-workspace/minimal/.scala-build/minimal_f4dd477a3a/classes/main/",
    "analysisOut": "file:///Users/ckipp/Documents/scala-workspace/minimal/.scala-build/.bloop/minimal_f4dd477a3a/minimal_f4dd477a3a-analysis.bin"
  }
}


[Trace - 06:01:07 PM] Received response 'buildTarget/compile - (19)' in 140ms
Result: {
  "originId": "METALS-$f5924e8b-0e59-4300-8cfc-9173bb99a6b7",
  "statusCode": 1
}
Error: null
```
In this case everything is already fully compiled. So the compile request comes in, and you see what is posted above.  In this scenario, do you think it just shouldn't send any `build/taskStart`?

**Reply by tgodzik** [reply]
Maybe just the compilation report in that case? Would actually make sense, since no actual task was run.

**Reply by adpi2** [reply]
I agree it makes sense not to sent any `taskStart` and `taskEnd`. However it would be quite hard to implement in sbt because it is hard to know in advance if a compilation is going to be a no-op. So I suggest we keep the `noOp` field in `CompileReport` for now. Also we should probably add it to [CompileResult](https://github.com/build-server-protocol/build-server-protocol/blob/eac6038263510dad9f299e1bce3a02b6a272763b/spec/src/main/resources/META-INF/smithy/bsp/bsp.smithy#L1197-L1208) too.

**Reply by ckipp01** [reply]
> So I suggest we keep the noOp field in CompileReport for now

Yes, I agree. It makes sense to keep it. You saw the Bloop PR and I agree, it's not super well defined what should happen in the case of a no-op, but if we can avoid the task start/end in the situation. of a no-op it's probably worth doing.

In the case of sbt, if there is no way to know ahead of time, then I think we sort of just need to deal with it and it's fine.
